File: ttt.js
// ttt_test - Speed test for random games of tic tac toe
// Released into the public domain by Brian Chen (differental) and Jeremy Chen (intgrah), 2025.
//
// This is free and unencumbered software released into the public domain.
//
// Anyone is free to copy, modify, publish, use, compile, sell, or distribute
// this software, either in source code form or as a compiled binary, for any
// purpose, commercial or non-commercial, and by any means.
//
// In jurisdictions that recognize copyright laws, the author or authors of
// this software dedicate any and all copyright interest in the software to
// the public domain. We make this dedication for the benefit of the public at
// large and to the detriment of our heirs and successors. We intend this
// dedication to be an overt act of relinquishment in perpetuity of all
// present and future rights to this software under copyright law.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// For more information, please refer to <https://unlicense.org/>

const BOARD_SIZE = 20;
const BOARD_SIZE_SQUARED = BOARD_SIZE * BOARD_SIZE;
const WIN_CONDITION = 10;

class Board {
    constructor() {
        this.board = Array(BOARD_SIZE);
        for (let i = 0; i < BOARD_SIZE; i++) this.board[i] = Array(BOARD_SIZE);

        this.cols = Array(BOARD_SIZE);
        this.rows = Array(BOARD_SIZE);
        this.diag = Array(2 * BOARD_SIZE - 1);
        this.anti = Array(2 * BOARD_SIZE - 1);
    }

    clear() {
        for (let i = 0; i < BOARD_SIZE; i++) this.board[i].fill(false);
        this.cols.fill(0);
        this.rows.fill(0);
        this.diag.fill(0);
        this.anti.fill(0);
    }

    update(x, y) {
        this.board[y][x] = true;
        this.cols[x]++;
        this.rows[y]++;
        this.diag[x - y + BOARD_SIZE - 1]++;
        this.anti[x + y]++;
    }

    check_win(x, y) {
        // Column
        if (this.cols[x] >= WIN_CONDITION) {
            let acc = 1;
            for (let ya = y; ++ya < BOARD_SIZE && this.board[ya][x]; )
                if (++acc >= WIN_CONDITION) return true;
            for (let ya = y; 0 <= --ya && this.board[ya][x]; )
                if (++acc >= WIN_CONDITION) return true;
        }

        // Row
        if (this.rows[y] >= WIN_CONDITION) {
            let acc = 1;
            for (let xa = x; ++xa < BOARD_SIZE && this.board[y][xa]; )
                if (++acc >= WIN_CONDITION) return true;
            for (let xa = x; 0 <= --xa && this.board[y][xa]; )
                if (++acc >= WIN_CONDITION) return true;
        }

        // Diagonal
        if (this.diag[x - y + BOARD_SIZE - 1] >= WIN_CONDITION) {
            let acc = 1;
            for (
                let xa = x, ya = y;
                ++xa < BOARD_SIZE && ++ya < BOARD_SIZE && this.board[ya][xa];

            )
                if (++acc >= WIN_CONDITION) return true;
            for (
                let xa = x, ya = y;
                0 <= --xa && 0 <= --ya && this.board[ya][xa];

            )
                if (++acc >= WIN_CONDITION) return true;
        }

        // Anti-diagonal
        if (this.anti[x + y] >= WIN_CONDITION) {
            let acc = 1;
            for (
                let xa = x, ya = y;
                ++xa < BOARD_SIZE && 0 <= --ya && this.board[ya][xa];

            )
                if (++acc >= WIN_CONDITION) return true;
            for (
                let xa = x, ya = y;
                0 <= --xa && ++ya < BOARD_SIZE && this.board[ya][xa];

            )
                if (++acc >= WIN_CONDITION) return true;
        }
        return false;
    }
}

let x = 172916342314;
function rand() {
    // Xorshift
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    x &= 0xffffffff;
    return x;
}

function shuffle(array) {
    // Fisher-Yates
    for (let i = array.length - 1; i > 0; i--) {
        const m = i + 1;
        const j = Math.abs(rand()) % m;
        const temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

function do_game(free_cells, circle, cross) {
    for (let i = 0; i < BOARD_SIZE_SQUARED; i++) free_cells[i] = i;
    shuffle(free_cells);

    circle.clear();
    cross.clear();

    for (let i = 0; i < BOARD_SIZE_SQUARED; i++) {
        const x = free_cells[i] % BOARD_SIZE;
        const y = (free_cells[i] - x) / BOARD_SIZE;
        const player = i % 2;
        const board = player === 0 ? circle : cross;
        board.update(x, y);
        if (board.check_win(x, y)) return player;
    }

    return -1;
}

function main() {
    const n = 10000;
    let o = 0;
    let x = 0;
    let draw = 0;

    const free_cells = Array(BOARD_SIZE_SQUARED);
    const circle = new Board();
    const cross = new Board();

    const start = performance.now();

    for (let i = 0; i < n; i++) {
        const res = do_game(free_cells, circle, cross);
        switch (res) {
            case 0:
                o += 1;
                break;
            case 1:
                x += 1;
                break;
            case -1:
                draw += 1;
                break;
        }
    }

    const end = performance.now();

    console.log(`O/X/Draw: ${o}/${x}/${draw}`);
    console.log(`Time taken: ${Math.round(end - start)} ms`);
}

main();

File: ttt.py
# ttt_test - Speed test for random games of tic tac toe
# Released into the public domain by Brian Chen (differental) and Jeremy Chen (intgrah), 2025.
#
# This is free and unencumbered software released into the public domain.
#
# Anyone is free to copy, modify, publish, use, compile, sell, or distribute
# this software, either in source code form or as a compiled binary, for any
# purpose, commercial or non-commercial, and by any means.
#
# In jurisdictions that recognize copyright laws, the author or authors of
# this software dedicate any and all copyright interest in the software to
# the public domain. We make this dedication for the benefit of the public at
# large and to the detriment of our heirs and successors. We intend this
# dedication to be an overt act of relinquishment in perpetuity of all
# present and future rights to this software under copyright law.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# For more information, please refer to <https://unlicense.org/>

import random
import time

BOARD_SIZE = 20
BOARD_SIZE_SQUARED = BOARD_SIZE * BOARD_SIZE
WIN_CONDITION = 10


class Board:
    board: list[list[bool]]
    cols: list[int]
    rows: list[int]
    diag: list[int]
    anti: list[int]

    def __init__(self):
        self.board = [[False] * BOARD_SIZE for _ in range(BOARD_SIZE)]
        self.cols = [0] * BOARD_SIZE
        self.rows = [0] * BOARD_SIZE
        self.diag = [0] * (2 * BOARD_SIZE - 1)
        self.anti = [0] * (2 * BOARD_SIZE - 1)

    def update(self, x, y):
        self.board[y][x] = True
        self.cols[x] += 1
        self.rows[y] += 1
        self.diag[x - y + BOARD_SIZE - 1] += 1
        self.anti[x + y] += 1

    def check_win(self, x, y) -> bool:
        # Column
        if self.cols[x] >= WIN_CONDITION:
            acc = 1
            ya = y + 1
            while ya < BOARD_SIZE and self.board[ya][x]:
                acc += 1
                ya += 1
            ya = y - 1
            while 0 <= ya and self.board[ya][x]:
                acc += 1
                ya -= 1
            if acc >= WIN_CONDITION:
                return True

        # Row
        if self.rows[y] >= WIN_CONDITION:
            acc = 1
            xa = x + 1
            while xa < BOARD_SIZE and self.board[y][xa]:
                acc += 1
                xa += 1
            xa = x - 1
            while 0 <= xa and self.board[y][xa]:
                acc += 1
                xa -= 1
            if acc >= WIN_CONDITION:
                return True

        # Diagonal
        if self.diag[x - y + BOARD_SIZE - 1] >= WIN_CONDITION:
            acc = 1
            xa = x + 1
            ya = y + 1
            while xa < BOARD_SIZE and ya < BOARD_SIZE and self.board[ya][xa]:
                acc += 1
                xa += 1
                ya += 1
            xa = x - 1
            ya = y - 1
            while 0 <= xa and 0 <= ya and self.board[ya][xa]:
                acc += 1
                xa -= 1
                ya -= 1
            if acc >= WIN_CONDITION:
                return True

        # Anti-diagonal
        if self.anti[x + y] >= WIN_CONDITION:
            acc = 1
            xa = x + 1
            ya = y - 1
            while xa < BOARD_SIZE and 0 <= ya and self.board[ya][xa]:
                acc += 1
                xa += 1
                ya -= 1
            xa = x - 1
            ya = y + 1
            while 0 <= xa and ya < BOARD_SIZE and self.board[ya][xa]:
                acc += 1
                xa -= 1
                ya += 1
            if acc >= WIN_CONDITION:
                return True

        return False


def do_game() -> int:
    free_cells: list[tuple[int, int]] = [
        (x, y) for y in range(BOARD_SIZE) for x in range(BOARD_SIZE)
    ]
    random.shuffle(free_cells)

    circle = Board()
    cross = Board()

    for i, (x, y) in enumerate(free_cells):
        if i % 2 == 0:
            circle.update(x, y)
            if circle.check_win(x, y):
                return 0
        else:
            cross.update(x, y)
            if cross.check_win(x, y):
                return 1

    return -1


def main() -> None:
    random.seed(1729)

    n = 10000
    o = 0
    x = 0
    draw = 0

    start: float = time.time()

    for _ in range(n):
        res: int = do_game()
        if res == 0:
            o += 1
        elif res == 1:
            x += 1
        else:
            draw += 1

    end: float = time.time()
    print(f"O/X/Draw: {o}/{x}/{draw}")
    print(f"Time taken: {(end - start) * 1000:.2f} ms")


if __name__ == "__main__":
    main()
